pkg/oci/layout/image.go

```diff
@@ -0,0 +1,103 @@
+//
+// Copyright 2021 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package layout
+
+import (
+	"fmt"
+
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/layout"
+	"github.com/google/go-containerregistry/pkg/v1/partial"
+	"github.com/sigstore/cosign/pkg/oci"
+	"github.com/sigstore/cosign/pkg/oci/siglayer"
+)
+
+// SignedImage provides access to a remote image reference, and its signatures.
+func SignedImage(path string) (oci.SignedImage, error) {
+	p, err := layout.FromPath(imagePath(path))
+	if err != nil {
+		return nil, err
+	}
+	img, err := p.Image(v1.Hash{})
+	if err != nil {
+		return nil, err
+	}
+
+	return &image{
+		Image: img,
+		path:  path,
+	}, nil
+}
+
+type image struct {
+	path string
+	v1.Image
+}
+
+var _ oci.SignedImage = (*image)(nil)
+
+type sigs struct {
+	v1.Image
+}
+
+var _ oci.Signatures = (*sigs)(nil)
+
+// Get implements oci.Signatures
+func (s *sigs) Get() ([]oci.Signature, error) {
+	layers, err := s.Image.Layers()
+	if err != nil {
+		return nil, err
+	}
+	var signatures []oci.Signature
+	for _, l := range layers {
```

Comment:
I'd rework the logic around this to fetch the `v1.Manifest`, walk the descriptors, and for each of the descriptors lookup the layer by digest (in the descriptor).  This is a weird use of `partial` and I'm not sure it WAI.